### PR TITLE
Fix data generation for custom dynamic registry

### LINKED ISSUE #3215: [FAPI 0.86] RegistryLoaderMixin is overwriting Registry Path
This mixing is changing the path for every, non vanilla Registry. Even if a Registry is not managed  with `DynamicRegistries`:

https://github.com/FabricMC/fabric/blob/e3d1b5f3d89f780b644f6e967963224c2cdef194/fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/RegistryLoaderMixin.java#L65-L71

The results in registries not getting loaded (quiqueck/BetterEnd#273)

### PR DESCRIPTION
**Pull request by apple502j**
Fixes #3215

This also adds a test for the dynamic registry provider.

### CODE REVIEW

**Comment by Technici4n** [comment]
Should this only be done for registries registered via the API?

**Comment by apple502j** [comment]
@Technici4n Perhaps. Another solution is to make a new method (with the check you mentioned in place), which providers can override.

File: fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestContent.java [hunk 1]
```diff
@@ -46,6 +50,14 @@ public class DataGeneratorTestContent implements ModInitializer {
 
 	public static final RegistryKey<ItemGroup> SIMPLE_ITEM_GROUP = RegistryKey.of(RegistryKeys.ITEM_GROUP, new Identifier(MOD_ID, "simple"));
 
+	public static final RegistryKey<Registry<TestDatagenObject>> TEST_DATAGEN_DYNAMIC_REGISTRY_KEY =
+			RegistryKey.ofRegistry(new Identifier("fabric", "test_datagen_dynamic"));
+	public static final RegistryKey<TestDatagenObject> TEST_DYNAMIC_REGISTRY_ITEM_KEY = RegistryKey.of(
+			TEST_DATAGEN_DYNAMIC_REGISTRY_KEY,
+			new Identifier(MOD_ID, "tiny_potato")
+	);
+	public static final TestDatagenObject TEST_DYNAMIC_REGISTRY_ITEM = new TestDatagenObject(":tiny_potato:");
```
Comment (modmuss50) [inline fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestContent.java]: I dont think this follows the correct pattern (Fabric API should set a good example), this should only be created in the datagenerator's buildRegistry.

File: fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java [hunk 2]
```diff
@@ -114,6 +119,14 @@ public void onInitializeDataGenerator(FabricDataGenerator dataGenerator) {
 		}
 	}
 
+	@Override
+	public void buildRegistry(RegistryBuilder registryBuilder) {
+		registryBuilder.addRegistry(
+				DataGeneratorTestContent.TEST_DATAGEN_DYNAMIC_REGISTRY_KEY,
+				(registerable) -> { }
```
Comment (modmuss50) [inline fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java]: I would create a full method for this and register the test entry to this `Registerable` see TestBiomes.bootstrap()

File: fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java [hunk 3]
```diff
@@ -364,4 +377,24 @@ public void accept(BiConsumer<Identifier, LootTable.Builder> consumer) {
...
+	 * Note that Biome API testmod provides the test for vanilla dynamic registries.
+	 */
+	private static class TestDynamicRegistryProvider extends FabricDynamicRegistryProvider {
+		TestDynamicRegistryProvider(FabricDataOutput output, CompletableFuture<RegistryWrapper.WrapperLookup> registriesFuture) {
+			super(output, registriesFuture);
+		}
+
+		@Override
+		protected void configure(RegistryWrapper.WrapperLookup registries, Entries entries) {
+			entries.add(TEST_DYNAMIC_REGISTRY_ITEM_KEY, TEST_DYNAMIC_REGISTRY_ITEM);
```
Comment (modmuss50) [inline fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java]: `TEST_DYNAMIC_REGISTRY_ITEM` should be retrived via the WrapperLookup and not have a constant.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
The code changes are fine, however I think we should really demonstate the correct pattern for the testmod.

**Review by modmuss50 — APPROVED** [review]

**Comment by modmuss50** [comment]
> Should this only be done for registries registered via the API?

Yeah, I think this would be wise, and should match what is done in https://github.com/FabricMC/fabric/pull/3219

I will merge 3219 first, and then update this to use the stuff added in this PR.

**Pull request merged**